File: Qytetet_el_juego/p3/QytetetJava/src/modeloqytetet/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author mati
 */
public class Casilla {
    
    private int numeroCasilla;
    private int coste;
    private int numHoteles;
    private int numCasas;
    private TipoCasilla tipo; //los de tipo calle tienen hoteles y casas, el resto no?
    private TituloPropiedad titulo;

    public Casilla(int numeroCasilla, int coste, int numHoteles, int numCasas, TipoCasilla tipo, TituloPropiedad titulo) {
        this.numeroCasilla = numeroCasilla;
        this.coste = coste;
        this.numHoteles = numHoteles;
        this.numCasas = numCasas;
        this.tipo = tipo;
        this.titulo = titulo;
        this.titulo.setCasilla(this);
        
    }

    public Casilla(int numeroCasilla, int coste, TipoCasilla tipo) {
        this.numeroCasilla = numeroCasilla;
        this.coste = coste;
        this.tipo = tipo;
        this.titulo = null;
        this.numHoteles = 0;
        this.numCasas = 0;
    }

    public TipoCasilla getTipo() {
        return this.tipo;
    }
    
    public TituloPropiedad getTituloPropiedad() {
        return titulo;
    }
    
    
    TituloPropiedad asignarPropietario( Jugador jugador){
        this.titulo.setPropietario(jugador);
        return titulo ;
    }
    
    int calcularValorHipoteca(){
        int cantidad = (this.titulo.getHipotecaBase()+ (this.numCasas * 5/10 * this.titulo.getHipotecaBase()+ (this.numHoteles * this.titulo.getAlquilerBase())));
        return cantidad ;
    }
    
    int cancelarHipoteca(){ //***********************
        return (this.calcularValorHipoteca()*10/100) + this.calcularValorHipoteca();
    }
    
    int cobrarAlquiler(){ 
        int costeAlquilerBase = this.titulo.getAlquilerBase();
        int coste = (costeAlquilerBase + (2 *((this.numCasas * 10/5) + (this.numHoteles * 2)))) ;
        this.titulo.cobrarAlquiler(coste);
        return coste;
    }
    
    int edificarCasa(){
        this.setNumCasas(this.numCasas + 1);
        int costeEdificarCasa = this.getPrecioEdificar();
        return costeEdificarCasa ;
    }
    
    int edificarHotel(){
        this.setNumHoteles(numHoteles + 1 );
        int costeEdificarHotel = this.getPrecioEdificar();
        return costeEdificarHotel ;
    }
    
    boolean estaHipotecada(){
        return this.titulo.getHipotecada();
    } 
    
    
    int getCoste() {
        return coste;
    }
    
   
    int getCosteHipoteca(){
        return this.calcularValorHipoteca();
    }
    
    
    public int getNumeroCasilla() { //package que pongo a publio *********************+
        return numeroCasilla;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    int getNumCasas() {
        return numCasas;
    }
    
    
    int getPrecioEdificar(){
        return this.titulo.getPrecioEdificar() ;
    }
   
    int hipotecar(){
        this.titulo.setHipotecada(true);
        return this.getCosteHipoteca();
    }
    
    int precioTotalComprar(){
        return this.titulo.getAlquilerBase();//////////////////////////////////
    }
   
    boolean propietarioEncarcelado(){
        return this.titulo.propietaroEncarcelado();
    }
    
    boolean sePuedeEdificarCasa(){
        boolean sepuede = false ;
        if(this.numCasas < 4 && this.soyEdificable()){
            sepuede = true ;
        }
        return sepuede ;
    }
    
    boolean sePuedeEdificarHotel(){
        boolean sepuede = false ;
        if(this.numHoteles < 4 && soyEdificable() ){
            sepuede = true ;
        }
        return sepuede ;
    }
    

    void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }

    void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }
    
    
    //Devuelve cierto sólo si es una casilla de tipo CALLE.
    boolean soyEdificable(){
        return this.tipo == TipoCasilla.CALLE;
    }
    
    public boolean tengoPropietario(){ // cambio de package a publico
        return this.titulo.tengoPropietario();
    }
   
    int venderTitulo(){
        int precio = this.precioTotalComprar();
        precio += this.numHoteles * this.getPrecioEdificar();
        precio += this.numCasas * this.getPrecioEdificar();
        precio *= this.titulo.getFactorDevolucion(); // tengo un float aqui.......................................
        this.titulo.setPropietario(null);
        //this.titulo = null ;
        this.setNumCasas(0);//this.numCasas = 0 ;
        this.setNumHoteles(0);//this.numHoteles = 0;
        return precio ;
    }
    
    

    private void setTituloPropiedad(TituloPropiedad titulo) {
        this.titulo = titulo;
    }
    
   
    private void asignarTituloPropiedad(){
        this.titulo = this.getTituloPropiedad() ;
    }
    

    @Override
    public String toString() {
        return "Casilla{" + "numeroCasilla=" + numeroCasilla + ",\n coste=" + coste +
                ",\n numHoteles=" + numHoteles + ",\n numCasas=" + numCasas +
                ",\n tipo Casilla=" + tipo + ",\n titulo en propiedad=" + titulo + '}';
    }
    
    
    
}

File: Qytetet_el_juego/p3/QytetetRuby/lib/casilla.rb
 #encoding: utf-8
 # To change this license header, choose License Headers in Project Properties.
 # To change this template file, choose Tools | Templates
 # and open the template in the editor.
 
 require_relative "tipo_casilla"
 require_relative "titulo_propiedad"

module ModeloQytetet
  class Casilla
    
    attr_accessor :numeroCasilla, :coste, :numHoteles, :numCasas, :tipo, :titulo
       
    def initialize(numCasilla,cos,numH,numCasa, tip, titul)  # dos constructores
      @numeroCasilla = numCasilla  # int
      @coste = cos  # int, coste de compra del titulo de propiedad de la casilla
      @numHoteles = numH  # int
      @numCasas = numCasa  # int
      @tipo = tip  # TipoCasilla
      @titulo = titul  # TituloPropiedad
      if @titulo != nil 
        @titulo.setCasilla(self) 
      end
      
    end
    
    def self.crear_casillas_calle(numCasilla,cos, titul)
      new(numCasilla,cos, 0, 0, TipoCasilla::CALLE, titul)
    end
    
    def self.crear_casillas_especiales(numCasilla, tip)
      new(numCasilla,0, 0, 0, tip, nil)
    end
    
    def asignarPropietario(jugador) #TituloPropiedad asignarPropietario( Jugador jugador)
      @titulo.setPropietario(jugador)
    end
    
    def calcularValorHipoteca #int calcularValorHipoteca()
      return (@titulo.hipotecaBase + 2 * (@numCasas * 5/10 * @titulo.hipotecaBase + @numHoteles + @titulo.hipotecaBase ))
    end
    
    def cancelarHipoteca #int cancelarHipoteca() #***************************************************************  @titulo.hipotecada = false
      return (calcularValorHipoteca()*10/100) + calcularValorHipoteca();
    end
    
    def cobrarAlquiler #int cobrarAlquiler()
      costeAlquilerBase = @titulo.alquilerBase
      coste = (costeAlquilerBase + (2 *((@numCasas * 5/10) + (@numHoteles * 2))))
      @titulo.cobrarAlquiler(coste)
      return coste
    end
    
    def edificarCasa #int edificarCasa()
      @numCasas += 1
      return @titulo.precioEdificar
    end
    
    def edificarHotel #int edificarHotel()
      @numHoteles += 1
      return @titulo.precioEdificar
    end
    
    def estaHipotecada #boolean estaHipotecada()
      return @titulo.hipotecada
    end
    
    def getPrecioEdificar #int getPrecioEdificar()
      @titulo.precioEdificar
    end
    
    def hipotecar #int hipotecar()
      @titulo.hipotecada=true ##############################
      cantidadRecibida = calcularValorHipoteca
      return cantidadRecibida
    end
    
    def precioTotalComprar #int precioTotalComprar()
      @titulo.alquilerBase
    end
    
    def propietarioEncarcelado #boolean propietarioEncarcelado()
      @titulo.propietaroEncarcelado
    end
    
    def sePuedeEdificarCasa #boolean sePuedeEdificarCasa()
      return (@numCasas < 4)
    end
    
    def sePuedeEdificarHotel #boolean sePuedeEdificarHotel()
      return (@numHoteles < 4)
    end
    
    def soyEdificable #boolean soyEdificable()
      return (@tipo == TipoCasilla::CALLE)
    end
    
    def tengoPropietario #boolean tengoPropietario() 
      @titulo.tengoPropietario
    end
    
    def venderTitulo #int venderTitulo()
      precio = precioTotalComprar
      precio += @numHoteles * getPrecioEdificar
      precio += @numCasas * getPrecioEdificar
      precio *= @titulo.factorRevalorizacion
      asignarPropietario(nil)
      @numHoteles = 0
      @numCasas = 0
      return precio
    end
    
    private
    
    def asignarTituloPropiedad #private void asignarTituloPropiedad()
      raise ArgumentError,'metodo sin implementar'
    end
    
    public
    
    def to_s
      cadena ="numeroCasilla: #{@numeroCasilla} \n 
      coste: #{@coste} \n 
      numHoteles: #{@numHoteles} \n
      numCasas: #{@numCasas} \n 
      tipo: #{@tipo} \n "
      if(@titulo != nil)
        cadena +="titulo: #{@titulo.to_s} "
      else
        cadena +="titulo: no tiene titulo"
      end
      return cadena
    end 
    
  end
end
